### PR TITLE
Add integration test cases for report definition rest APIs

### PR DESCRIPTION
**Pull request by joshuali925**
*Issue #, if available:*

*Description of changes:*
- Add integration test cases for report definition rest APIs
- Allow custom error types in validateErrorResponse helper function

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                    "format":{
+                        "duration":"PT1H",
+                        "fileFormat":"Pdf",
+                        "limit":1000,
+                        "header":"optional header",
+                        "footer":"optional footer"
+                    },
+                    $trigger
+                    "delivery":{
+                        "recipients":["banantha@amazon.com"],
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: may be use nobody@email.com for testing?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        )
+        reportDefinitionResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/definition",
+            reportDefinitionRequest,
+            RestStatus.OK.status
+        )
+        reportDefinitionId = reportDefinitionResponse.get("reportDefinitionId").asString
+        Assert.assertNotNull("reportDefinitionId should be generated", reportDefinitionId)
+        Thread.sleep(3000)
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: I am not sure if we require 3 seconds

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+            "",
+            RestStatus.OK.status
+        )
+        Assert.assertEquals(
+            reportDefinitionId,
+            reportDefinitionResponse.get("reportDefinitionDetails").asJsonObject.get("id").asString
+        )
+        Assert.assertEquals(
+            newName,
+            reportDefinitionResponse.get("reportDefinitionDetails").asJsonObject.get("reportDefinition").asJsonObject.get(
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: may be indent with second .get() in new line?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            "$BASE_REPORTS_URI/definition/$reportDefinitionId",
+            "",
+            RestStatus.NOT_FOUND.status
+        )
+        validateErrorResponse(reportDefinitionResponse, RestStatus.NOT_FOUND.status)
+        Thread.sleep(100)
+    }
+
+    fun `test invalid get, update, delete report definition`() {
+        var reportDefinitionResponse = executeRequest(
```
Comment (akbhatta) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: since no report definitions created this test may not cover some edge condition. may be we can create few report definitions before testing invalid ids?

**Review by akbhatta — COMMENTED** [review]
May be other values can be validated in some test cases?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt (hunk 1 above)
Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: good idea.. updated

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt (hunk 2 above)
Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: changed to 1 second

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt (hunk 3 above)
Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: updated

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt (hunk 4 above)
Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: merged into test function above so there are existing report definitions

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+        """.trimIndent()
+        reportDefinitionRequest = constructReportDefinitionRequest(trigger)
+        reportDefinitionResponse = executeRequest(
+            RestRequest.Method.POST.name,
+            "$BASE_REPORTS_URI/definition",
+            reportDefinitionRequest,
+            RestStatus.BAD_REQUEST.status
+        )
+        validateErrorResponse(reportDefinitionResponse, RestStatus.BAD_REQUEST.status, "illegal_argument_exception")
+        Thread.sleep(100)
```
Comment (davidcui-amzn) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt:298]: What's the reason for adding the `Thread.sleep`?

File: reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                        "channelIds":["optional_channelIds"]
+                    }
+                }
+            }
+            """.trimIndent()
+    }
+
+    fun `test create, get, update, delete report definition`() {
+        var reportDefinitionRequest = constructReportDefinitionRequest()
+        var reportDefinitionResponse = executeRequest(
```
Comment (davidcui1225) [inline reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: Should we make these `val` vs. `var` so that they are immutable?

Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: I reused those variables for other requests, won't work if they are immutable

Comment (akbhatta) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: I think it is better to use val and use another variable if they are not related.

Comment (joshuali925) [reply reports-scheduler/src/test/kotlin/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/ReportDefinitionIT.kt]: will refactor, thanks

**Review by davidcui1225 — APPROVED** [review]
LGTM

**Review by akbhatta — APPROVED** [review]
